File: app/src/main/java/ua/in/quireg/anothermovieapp/movieList/MainActivity.java
package ua.in.quireg.anothermovieapp.movieList;

import android.content.Context;
import android.content.Intent;
import android.content.res.TypedArray;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.view.OnApplyWindowInsetsListener;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.viewpager.widget.ViewPager;

import com.google.android.material.appbar.AppBarLayout;
import com.google.android.material.tabs.TabLayout;

import ua.in.quireg.anothermovieapp.R;
import ua.in.quireg.anothermovieapp.common.Constants;
import ua.in.quireg.anothermovieapp.common.MovieItem;
import ua.in.quireg.anothermovieapp.interfaces.OnFragmentInteractionListener;
import ua.in.quireg.anothermovieapp.movieDetails.MovieActivity;

public class MainActivity extends AppCompatActivity implements OnFragmentInteractionListener {

    private static final String LOG_TAG = MainActivity.class.getSimpleName();

    private static final int TOP_RATED_MOVIES_TAB_POSITION = 0;
    private static final int FAVOURITE_MOVIES_TAB_POSITION = 1;
    private static final int POPULAR_MOVIES_TAB_POSITION = 2;

    private SectionsPagerAdapter mSectionsPagerAdapter;

    private int mSystemInsetTop = -1;
    private int mSystemInsetBottom = -1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        View rootView = LayoutInflater.from(this).inflate(R.layout.activity_ama_main_activity, null);
        setContentView(rootView);
        rootView.setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION);
        final Toolbar toolbar = findViewById(R.id.toolbar);
        toolbar.setTitleTextColor(getResources().getColor(R.color.primaryColor));
        setSupportActionBar(toolbar);
        final AppBarLayout appBarLayout = findViewById(R.id.appbar);

        mSectionsPagerAdapter =
                new SectionsPagerAdapter(getSupportFragmentManager(), this);

        // Set up the ViewPager with the sections adapter.
        ViewPager viewPager = findViewById(R.id.container);
        viewPager.setAdapter(mSectionsPagerAdapter);

        TabLayout tabLayout = findViewById(R.id.tabs);
        tabLayout.setupWithViewPager(viewPager);
        ViewCompat.setOnApplyWindowInsetsListener(rootView, (v, insets) -> {
            appBarLayout.setPadding(
                    v.getPaddingLeft(),
                    insets.getSystemWindowInsetTop(),
                    v.getPaddingRight(),
                    v.getPaddingBottom());


            mSystemInsetTop = insets.getSystemWindowInsetTop();
            mSystemInsetBottom = insets.getSystemWindowInsetBottom();

            TextView pageNumber = findViewById(R.id.pageNumberAndTotal);
            FrameLayout.LayoutParams lp = (FrameLayout.LayoutParams) pageNumber.getLayoutParams();
            lp.bottomMargin += insets.getSystemWindowInsetBottom();
            pageNumber.setLayoutParams(lp);

            ProgressBar pb = findViewById(R.id.progressBar);
            FrameLayout.LayoutParams lp2 = (FrameLayout.LayoutParams) pageNumber.getLayoutParams();
            lp.bottomMargin += insets.getSystemWindowInsetBottom();
            pb.setLayoutParams(lp2);

            return insets.consumeSystemWindowInsets();
        });
    }

    @Override
    public void onAttachFragment(@NonNull Fragment fragment) {
        super.onAttachFragment(fragment);
        ViewPager viewPager = findViewById(R.id.container);
        viewPager.setCurrentItem(FAVOURITE_MOVIES_TAB_POSITION);
        mSectionsPagerAdapter.getItem(FAVOURITE_MOVIES_TAB_POSITION).setUserVisibleHint(true);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_ama_main_activity, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        } else if (id == R.id.action_refresh) {
            SyncMovieService.startFetchMovies(this, Constants.POPULAR, "1");
            SyncMovieService.startFetchMovies(this, Constants.TOP_RATED, "1");
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onFragmentMessage(String TAG, Object data) {
        Intent i = new Intent(MainActivity.this, MovieActivity.class);
        Bundle args = new Bundle();
        args.putSerializable(Constants.MOVIE, (MovieItem) data);
        i.putExtras(args);
        startActivity(i);
    }

    /**
     * A {@link FragmentPagerAdapter} that returns a fragment corresponding to
     * one of the sections/tabs/pages.
     */
    public class SectionsPagerAdapter extends FragmentPagerAdapter {

        private Context mContext;
        private MoviesGridViewFragment popularMoviesFragment;
        private MoviesGridViewFragment topRatedMoviesFragment;
        private MoviesGridViewFragment favouritesMoviesFragment;

        SectionsPagerAdapter(FragmentManager fm, Context context) {
            super(fm);
            this.mContext = context;
        }

        @NonNull
        @Override
        public Fragment getItem(int position) {
            switch (position) {
                case POPULAR_MOVIES_TAB_POSITION:
                    if (popularMoviesFragment == null) {
                        popularMoviesFragment = new PopularMoviesGridViewFragment();
                        Bundle args = new Bundle();
                        args.putSerializable(Constants.FRAGMENT_TAG, Constants.POPULAR);
                        args.putSerializable(Constants.INSET_TOP, mSystemInsetTop);
                        args.putSerializable(Constants.INSET_BOT, mSystemInsetBottom);
                        popularMoviesFragment.setArguments(args);
                    }
                    return popularMoviesFragment;
                case TOP_RATED_MOVIES_TAB_POSITION:
                    if (topRatedMoviesFragment == null) {
                        topRatedMoviesFragment = new TopRatedMoviesGridViewFragment();
                        Bundle args = new Bundle();
                        args.putSerializable(Constants.FRAGMENT_TAG, Constants.TOP_RATED);
                        args.putSerializable(Constants.INSET_TOP, mSystemInsetTop);
                        args.putSerializable(Constants.INSET_BOT, mSystemInsetBottom);
                        topRatedMoviesFragment.setArguments(args);
                    }
                    return topRatedMoviesFragment;
                case FAVOURITE_MOVIES_TAB_POSITION:
                default:
                    if (favouritesMoviesFragment == null) {
                        favouritesMoviesFragment = new FavouritesMoviesGridViewFragment();
                        Bundle args = new Bundle();
                        args.putSerializable(Constants.FRAGMENT_TAG, Constants.FAVOURITES);
                        args.putSerializable(Constants.INSET_TOP, mSystemInsetTop);
                        args.putSerializable(Constants.INSET_BOT, mSystemInsetBottom);
                        favouritesMoviesFragment.setArguments(args);
                    }
                    return favouritesMoviesFragment;
            }
        }

        @Override
        public int getCount() {
            return 3;
        }

        @Override
        public CharSequence getPageTitle(int position) {
            switch (position) {
                case POPULAR_MOVIES_TAB_POSITION:
                    return mContext.getResources().getString(R.string.popular_tab_name);
                case TOP_RATED_MOVIES_TAB_POSITION:
                    return mContext.getResources().getString(R.string.top_rated_tab_name);
                case FAVOURITE_MOVIES_TAB_POSITION:
                default:
                    return mContext.getResources().getString(R.string.favourites_tab_name);
            }
        }
    }
}

File: app/src/main/java/com/anothermovieapp/movieList/MainActivity.kt
package com.anothermovieapp.movieList

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.viewpager.widget.ViewPager
import com.anothermovieapp.R
import com.anothermovieapp.common.Constants
import com.anothermovieapp.interfaces.OnFragmentInteractionListener
import com.anothermovieapp.movieDetails.MovieActivity
import com.anothermovieapp.repository.EntityDBMovie
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.tabs.TabLayout
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class MainActivity : AppCompatActivity() {
    lateinit var mSectionsPagerAdapter: SectionsPagerAdapter
    var popularMoviesFragment: MoviesGridViewFragment? = null
    var topRatedMoviesFragment: MoviesGridViewFragment? = null
    var favouritesMoviesFragment: MoviesGridViewFragment? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val rootView = LayoutInflater.from(this)
            .inflate(R.layout.activity_ama_main_activity, null)

        setContentView(rootView)
//        rootView.systemUiVisibility = (View.SYSTEM_UI_FLAG_LAYOUT_STABLE
//                or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
//                or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        toolbar.setTitleTextColor(resources.getColor(R.color.primaryColor, theme))
        setSupportActionBar(toolbar)
        val appBarLayout = findViewById<AppBarLayout>(R.id.appbar)
        // Set up the ViewPager with the sections adapter.
        val viewPager = findViewById<ViewPager>(R.id.container)
        popularMoviesFragment =
            supportFragmentManager.findFragmentByTag(
                "android:switcher:" + viewPager.id + ":" + POPULAR_MOVIES_TAB_POSITION)
                    as MoviesGridViewFragment?
        if (popularMoviesFragment == null) {
            popularMoviesFragment = PopularMoviesGridViewFragment()
        }
        popularMoviesFragment!!.fragmentListener(fragmentListener)
        topRatedMoviesFragment =
            supportFragmentManager.findFragmentByTag(
                "android:switcher:" + viewPager.id + ":" + TOP_RATED_MOVIES_TAB_POSITION)
                    as MoviesGridViewFragment?
        if (topRatedMoviesFragment == null) {
            topRatedMoviesFragment = TopRatedMoviesGridViewFragment()
        }
        topRatedMoviesFragment!!.fragmentListener(fragmentListener)
        favouritesMoviesFragment =
            supportFragmentManager.findFragmentByTag(
                "android:switcher:" + viewPager.id + ":" + FAVOURITE_MOVIES_TAB_POSITION)
                    as MoviesGridViewFragment?
        if (favouritesMoviesFragment == null) {
            favouritesMoviesFragment = FavouritesMoviesGridViewFragment()
        }
        favouritesMoviesFragment!!.fragmentListener(fragmentListener)
        mSectionsPagerAdapter = SectionsPagerAdapter(supportFragmentManager)


        viewPager.adapter = mSectionsPagerAdapter
        val tabLayout = findViewById<TabLayout>(R.id.tabs)
        tabLayout.setupWithViewPager(viewPager)
//        ViewCompat.setOnApplyWindowInsetsListener(rootView) { v: View, insets: WindowInsetsCompat ->
//            ViewCompat.setOnApplyWindowInsetsListener(rootView, null)
//            appBarLayout.setPadding(
//                    v.paddingLeft,
//                    insets.systemWindowInsetTop,
//                    v.paddingRight,
//                    v.paddingBottom)
//            mSystemInsetTop = insets.systemWindowInsetTop
//            mSystemInsetBottom = insets.systemWindowInsetBottom
//            val pageNumber = findViewById<TextView>(R.id.pageNumberAndTotal)
//            val lp = pageNumber.layoutParams as FrameLayout.LayoutParams
//            lp.bottomMargin += insets.systemWindowInsetBottom
//            pageNumber.layoutParams = lp
//            val pb = findViewById<ProgressBar>(R.id.progressBar)
//            val lp2 = pageNumber.layoutParams as FrameLayout.LayoutParams
//            lp.bottomMargin += insets.systemWindowInsetBottom
//            pb.layoutParams = lp2
//            insets.consumeSystemWindowInsets()
//        }


    }

//    override fun onAttachFragment(fragment: Fragment) {
//        super.onAttachFragment(fragment)
//        val viewPager = findViewById<ViewPager>(R.id.container)
//        viewPager.currentItem = FAVOURITE_MOVIES_TAB_POSITION
//        mSectionsPagerAdapter!!.getItem(FAVOURITE_MOVIES_TAB_POSITION).userVisibleHint = true
//    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_ama_main_activity, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        val id = item.itemId
        if (id == R.id.action_settings) {
            return true
        } else if (id == R.id.action_refresh) {
            popularMoviesFragment?.reload()
            topRatedMoviesFragment?.reload()
            favouritesMoviesFragment?.reload()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    val fragmentListener = object : OnFragmentInteractionListener {
        override fun onFragmentMessage(TAG: String?, data: Any?) {
            val i = Intent(this@MainActivity, MovieActivity::class.java)
            val args = Bundle()
            args.putSerializable(Constants.MOVIE, data as EntityDBMovie?)
            i.putExtras(args)
            startActivity(i)
        }
    }

    /**
     * A [FragmentPagerAdapter] that returns a fragment corresponding to
     * one of the sections/tabs/pages.
     */
    inner class SectionsPagerAdapter internal constructor(
        fm: FragmentManager
    ) : FragmentPagerAdapter(fm, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {

        override fun getItem(position: Int): Fragment {
            return when (position) {
                POPULAR_MOVIES_TAB_POSITION -> {
//                    popularMoviesFragment = PopularMoviesGridViewFragment()
//                    val bundle = Bundle()
//                    bundle.putSerializable("listener", fragmentListener)
//                    popularMoviesFragment?.arguments = bundle
                    popularMoviesFragment!!
                }
                TOP_RATED_MOVIES_TAB_POSITION -> {
//                    topRatedMoviesFragment = TopRatedMoviesGridViewFragment()
//                    val bundle = Bundle()
//                    bundle.putSerializable("listener", fragmentListener)
//                    topRatedMoviesFragment?.arguments = bundle
                    topRatedMoviesFragment!!
                }
                else -> {
//                    favouritesMoviesFragment = FavouritesMoviesGridViewFragment()
//                    val bundle = Bundle()
//                    bundle.putSerializable("listener", fragmentListener)
//                    favouritesMoviesFragment?.arguments = bundle
                    favouritesMoviesFragment!!
                }
            }
        }

        override fun getCount(): Int {
            return 3
        }

//        override fun getPageTitle(position: Int): CharSequence {
//            return when (position) {
//                POPULAR_MOVIES_TAB_POSITION -> mContext.resources.getString(R.string.popular_tab_name)
//                TOP_RATED_MOVIES_TAB_POSITION -> mContext.resources.getString(R.string.top_rated_tab_name)
//                FAVOURITE_MOVIES_TAB_POSITION -> mContext.resources.getString(R.string.favourites_tab_name)
//                else -> mContext.resources.getString(R.string.favourites_tab_name)
//            }
//        }

    }

    companion object {
        private const val TOP_RATED_MOVIES_TAB_POSITION = 1
        private const val FAVOURITE_MOVIES_TAB_POSITION = 2
        private const val POPULAR_MOVIES_TAB_POSITION = 0
    }
}